src/levels/sparsetrianglelevels.jl

```diff
@@ -0,0 +1,185 @@
+end
+
+mutable struct VirtualSparseTriangleLevel
+    lvl
+    ex
+    Ti
+    I
+    J
+end
+function virtualize(ex, ::Type{SparseTriangleLevel{Ti, Lvl}}, ctx, tag=:lvl) where {Ti, Lvl}
+    sym = ctx.freshen(tag)
+    I = value(:($sym.I), Int)
+    J = value(:($sym.J), Int)
+    push!(ctx.preamble, quote
+        $sym = $ex
+    end)
+    lvl_2 = virtualize(:($sym.lvl), Lvl, ctx, sym)
+    VirtualSparseTriangleLevel(lvl_2, sym, Ti, I, J)
+end
+function (ctx::Finch.LowerJulia)(lvl::VirtualSparseTriangleLevel)
+    quote
+        $SparseTriangleLevel{$(lvl.Ti)}(
+            $(ctx(lvl.lvl)),
+            $(ctx(lvl.I)),
+            $(ctx(lvl.J)),
+        )
+    end
+end
+
+summary_f_code(lvl::VirtualSparseTriangleLevel) = "st($(summary_f_code(lvl.lvl)))"
+
+function virtual_level_size(lvl::VirtualSparseTriangleLevel, ctx)
+    ext_1 = Extent(literal(lvl.Ti(1)), lvl.I)
+    ext_2 = Extent(literal(lvl.Ti(1)), lvl.J)
+    (virtual_level_size(lvl.lvl, ctx)..., ext_1, ext_2)
+end
+
+function virtual_level_resize!(lvl::VirtualSparseTriangleLevel, ctx, dims...)
+    lvl.I = getstop(dims[end])
+    lvl.J = getstop(dims[end - 1])
+    lvl.lvl = virtual_level_resize!(lvl.lvl, ctx, dims[1:end-2]...)
+    lvl
+end
+
+virtual_level_eltype(lvl::VirtualSparseTriangleLevel) = virtual_level_eltype(lvl.lvl)
+virtual_level_default(lvl::VirtualSparseTriangleLevel) = virtual_level_default(lvl.lvl)
+
+function initialize_level!(lvl::VirtualSparseTriangleLevel, ctx::LowerJulia, pos)
+    lvl.lvl = initialize_level!(lvl.lvl, ctx, call(*, pos, lvl.I))
+    return lvl
+end
+
+function trim_level!(lvl::VirtualSparseTriangleLevel, ctx::LowerJulia, pos)
+    qos = ctx.freshen(:qos)
+    push!(ctx.preamble, quote
+        $qos = $(ctx(pos)) * $(ctx(lvl.I))
+    end)
+    lvl.lvl = trim_level!(lvl.lvl, ctx, value(qos))
+    return lvl
+end
+
+function assemble_level!(lvl::VirtualSparseTriangleLevel, ctx, pos_start, pos_stop)
+    qos_start = call(+, call(*, call(-, pos_start, lvl.Ti(1)), lvl.I), 1)
+    qos_stop = call(*, pos_stop, lvl.I)
+    assemble_level!(lvl.lvl, ctx, qos_start, qos_stop)
+end
+
+supports_reassembly(::VirtualSparseTriangleLevel) = true
+function reassemble_level!(lvl::VirtualSparseTriangleLevel, ctx, pos_start, pos_stop)
+    qos_start = call(+, call(*, call(-, pos_start, lvl.Ti(1)), lvl.I), 1)
+    qos_stop = call(*, pos_stop, lvl.I)
+    reassemble_level!(lvl.lvl, ctx, qos_start, qos_stop)
+    lvl
+end
+
+function freeze_level!(lvl::VirtualSparseTriangleLevel, ctx::LowerJulia, pos)
+    lvl.lvl = freeze_level!(lvl.lvl, ctx, call(*, pos, lvl.I))
+    return lvl
+end
+
+is_laminable_updater(lvl::VirtualSparseTriangleLevel, ctx, ::Union{Nothing, Laminate, Extrude}, protos...) =
+    is_laminable_updater(lvl.lvl, ctx, protos...)
+
+get_reader(fbr::VirtualSubFiber{VirtualSparseTriangleLevel}, ctx, ::Union{Nothing, Follow}, protos...) = get_readerupdater_triangular_dense_helper(fbr, ctx, get_reader, VirtualSubFiber, protos...)
```

Comment:
This line only consumes one protocol. It may be that the appropriate elementlevel get_reader function does not get called because this is passing a protocol to get_reader, but the element level version of get_reader does not take a protocol.